### ISSUE TITLE
React as Render Engine

### ISSUE THREAD

**Issue by andresgutgon** [bpampuch/pdfmake#1632]
### What is this?
Hi, first of all thanks for the Software 💪 
I'm using it to render a pdf on the command line from a JSON file, this is not important, the important part is that the lib works great it generates expect result. My concern is with DX (Development Experience). Doing a lot nested javascript objects is not funny. Here is my idea:

I saw this on Comming Soon:
```
build custom DSLs on top of document-definition-model (this is actually possible at the moment).
```

Maybe someone already said that but if instead a custom DSL this lib used a React Renderer. It would be awesome to do PDFs this way:
``` javascript 
  <Page>
     <Header title='Hello word'>
     <Table>...</Table>
  </Page>
```
Those component could receive `styles` so you can customize it. As I said maybe someone already said or even better it's already done 😄 . Just my 2 cents

**Comment by liborm85** [comment]
pdfmake uses JSON for document definition. Is not possible use React template engine.

You can make converter from React template definition to pdfmake json document definition.

**Comment by blikblum** [comment]
You may want to try https://react-pdf.org/ but there's no need to be React specific.

Is possible to configure JSX to call specific functions that would create the pdfmake definition.

Technically is doable, but the question is who will work on it?

I can do at most a proof of concept if someone is interested in picking up and covering the pdfmake options.

**Comment by andresgutgon** [comment]
Thanks @blikblum react-pdf looks cool! I already finished what I wanted with current pdfmake version (JSON). It was just curiosity, I don't want to port it to React :)

But this project works pretty well, React renderer it just would be a nice to have

**Comment by mtlewis** [comment]
Just stumbled on this issue - you might be interested in [jsx-pdf](http://github.com/schibsted/jsx-pdf/) which adds support for writing PDFs as jsx to pdfmake. Disclaimer: I'm one of the maintainers 😁

**Comment by danvim** [comment]
> Just stumbled on this issue - you might be interested in [jsx-pdf](https://github.com/schibsted/jsx-pdf/) which adds support for writing PDFs as jsx to pdfmake. Disclaimer: I'm one of the maintainers 😁

I've just started a new project [react-pdfmake-reconciler](https://github.com/danvim/react-pdfmake-reconciler) over the weekends and have been working on this these days. I then checked if anyone was talking about React in this repo and found this. I wasn't aware you had the same idea.

I originally also thought of making my own jsxFactories as you did, but in the end chose to use `react-reconciler`. This package may interoperate better with React (Context, hooks, devtools) and since it doesn't replace jsxFactories, it doesn't need TypeScript/babel plugin for the additional JSX conversion.
